onnxruntime/core/framework/allocation_planner.cc

```diff
@@ -742,9 +742,20 @@ class PlannerImpl {
             }
           }
         } else if (IsNonTensor(*node_output)) {
-          // we do not try sharing-optimization for non-tensors
-          AllocPlan(current).alloc_kind = AllocKind::kAllocate;
-          AllocPlan(current).program_counter.AddStart(program_counter);
+          // For optional type outputs, certain kernels allow re-using input OrtValues.
+          // Check if re-using is feasible.
+          if (!context_.IsParallelExecutionEnabled() &&
+              IsOptionalType(*node_output) &&
+              FindReusableInput(*pnode, static_cast<int>(output_arg_def_index), &reused)) {
+            // Reuse one of this node's input's as its output
+            Reuse(reused, current, AllocKind::kReuse);
+#if !defined(ORT_MINIMAL_BUILD) && defined(ORT_MEMORY_PROFILE)
+            InplaceReuse(reused, current);
+#endif
+          } else {  // we do not try sharing-optimization for other non-tensors
```

Comment:
Are there unit tests for this new logic?

Reply:
yes, there is a test to test the re-use of optional inputs as outputs.

Reply:
What about checks that there is no re-use of an optional output by a downstream node?

Reply:
Modified logic in allocation planner so that optional type inputs/outputs don't end up in free_list_